src/c/libp2p_c.cpp

```diff
@@ -0,0 +1,584 @@
+                          callback(nullptr, user_data);
+                        }
+                      });
+    return LIBP2P_SUCCESS;
+  } catch (...) {
+    return LIBP2P_ERROR_UNKNOWN;
+  }
+}
+
+// Utility functions implementation
+libp2p_multiaddr_t *libp2p_multiaddr_create(const char *addr_str) {
+  if (!addr_str) {
+    return nullptr;
+  }
+
+  try {
+    auto addr_result = libp2p::Multiaddress::create(addr_str);
+    if (!addr_result.has_value()) {
+      return nullptr;
+    }
+
+    return new libp2p_multiaddr_t(addr_result.value(), std::string(addr_str));
+  } catch (...) {
+    return nullptr;
+  }
+}
+
+void libp2p_multiaddr_destroy(libp2p_multiaddr_t *addr) {
+  delete addr;
+}
+
+const char *libp2p_multiaddr_to_string(libp2p_multiaddr_t *addr) {
+  if (!addr) {
+    return nullptr;
+  }
+  return addr->addr_str.c_str();
+}
+
+libp2p_peer_id_t *libp2p_peer_id_create(const char *base58_str) {
+  if (!base58_str) {
+    return nullptr;
+  }
+
+  try {
+    auto peer_id_result = libp2p::PeerId::fromBase58(base58_str);
+    if (!peer_id_result.has_value()) {
+      return nullptr;
+    }
+
+    return new libp2p_peer_id_t(peer_id_result.value(),
+                                std::string(base58_str));
+  } catch (...) {
+    return nullptr;
+  }
+}
+
+void libp2p_peer_id_destroy(libp2p_peer_id_t *peer_id) {
+  delete peer_id;
+}
+
+const char *libp2p_peer_id_to_string(libp2p_peer_id_t *peer_id) {
+  if (!peer_id) {
+    return nullptr;
+  }
+  return peer_id->peer_id_str.c_str();
+}
+
+// Error handling implementation
+const char *libp2p_error_string(libp2p_error_t error) {
+  switch (error) {
+    case LIBP2P_SUCCESS:
+      return "Success";
+    case LIBP2P_ERROR_INVALID_ARGUMENT:
+      return "Invalid argument";
+    case LIBP2P_ERROR_OUT_OF_MEMORY:
+      return "Out of memory";
+    case LIBP2P_ERROR_CONNECTION_FAILED:
+      return "Connection failed";
+    case LIBP2P_ERROR_PROTOCOL_ERROR:
+      return "Protocol error";
+    case LIBP2P_ERROR_IO_ERROR:
+      return "I/O error";
+    case LIBP2P_ERROR_TIMEOUT:
+      return "Timeout";
+    case LIBP2P_ERROR_NOT_FOUND:
+      return "Not found";
+    case LIBP2P_ERROR_ALREADY_EXISTS:
+      return "Already exists";
+    case LIBP2P_ERROR_UNKNOWN:
+      return "Unknown error";
+    default:
+      return "Invalid error code";
+  }
+}
+
+// Simplified logging implementation without spdlog dependency
+void libp2p_set_log_level(libp2p_log_level_t level) {
+  // For now, just a simplified implementation
+  // In a full implementation, you would integrate with the actual logging
+  // system
+  (void)level;  // Suppress unused parameter warning
+}
+
+const char *libp2p_host_peer_id(libp2p_host_t *host) {
+  if (!host || !host->host || !host->peer_id.has_value()) {
+    return nullptr;
+  }
+  static thread_local std::string id;
+  id = host->peer_id->toBase58();
+  return id.c_str();
+}
+
+const char *libp2p_host_default_listen(libp2p_host_t *host) {
```

Comment:
This method is ambiguous,
caller already knows listen address,
because he called `libp2p_host_listen`.
```suggestion
```